airbyte-integrations/connectors/source-postgres2/src/main/java/io/airbyte/integrations/source/postgres/DPostgresSource.java

```diff
@@ -143,29 +148,37 @@ public static void main(String[] args) throws IOException, InterruptedException
     // Run the engine asynchronously ...
     executor.execute(engine);
 
-    Iterator<JsonNode> delegateIterator = queue.iterator();
+    // retains the blocking behavior on hasNext (unlike Queue#iterator)
+    final Iterator<JsonNode> queueIterator = MoreStreams.toStream(queue).iterator();
 
-    return new AutoCloseableIterator<>() {
-      @Override
-      public void close() throws Exception {
-        engine.close();
-        executor.shutdown();
+    final AbstractIterator<JsonNode> iterator = new AbstractIterator<>() {
+      private boolean hasReachedLsn = false;
 
-        if(thrownError.get() != null) {
-          throw new RuntimeException(thrownError.get());
+      @Override
+      protected JsonNode computeNext() {
+        // if we have reached the lsn we stop, otherwise we have the potential to wait indefinitely for the next value.
+        if(!hasReachedLsn && queueIterator.hasNext()) {
+          final JsonNode next = queueIterator.next();
+          // todo fix this cast. the record passed to this iterator has to include the lsn somewhere. it can either be the full change event or some smaller object that just includes the lsn.
+          // we guarantee that this will always eventually return true, because we pick an LSN that already exists when we start the sync.
+          if(hasReachedLsnPredicate.test((ChangeEvent<String, String>) next)) {
```

Comment:
in the change event returned by debezium the lsn number _is_ present in `value.source.lsn`. so as long as that field gets propagated inside the iterator we are good. we just need to coordinate on how to do that.